Added ExecutePayloadV1 to engine_ namespace [the merge]

Giulio2002 opened the pull request:
PR Description:
* Added GetPayloadV1
* Added ExecutePayloadV1
* Added Tests for ExecutePayloadV1
* Separate calltracer from stagedsync (refactoring)
* Add GetPayloadV1, ExecutePayloadV1 and ForkchoiceUpdateV1 to RpcDaemon
* Compatible POS_FORKCHOICE_UPDATE API

Giulio2002 commented:
Ready for review

enriavil1 commented on eth/backend.go:
```diff
@@ -364,7 +370,8 @@ func New(stack *node.Node, config *ethconfig.Config, logger log.Logger) (*Ethere
 	}
-	ethBackendRPC := privateapi.NewEthBackendServer(ctx, backend, backend.chainDB, backend.notifications.Events, blockReader)
+	ethBackendRPC := privateapi.NewEthBackendServer(ctx, backend, backend.chainDB, backend.notifications.Events,
+		blockReader, chainConfig, backend.reverseDownloadCh, backend.statusCh)
```
Was this divided purposely into multiple lines?

AlexeyAkhunov commented on ethdb/privateapi/ethbackend.go:
```diff
@@ -147,3 +169,140 @@ func (s *EthBackendServer) Block(ctx context.Context, req *remote.BlockRequest)
+	// Check if current block is next for execution, if not, commission it and start
+	// Reverse-download the chain from its block number and hash.
+	if header.ParentHash != currentHead {
+		return &remote.EngineExecutePayloadReply{
```
If you leave here, no-one will take the value off the `s.statusCh`, and it may be incorrectly read by the next invocation

AlexeyAkhunov commented on ethdb/privateapi/ethbackend.go:
```diff
@@ -147,3 +179,105 @@ func (s *EthBackendServer) Block(ctx context.Context, req *remote.BlockRequest)
+		return nil, fmt.Errorf("not a proof-of-stake chain")
+	}
+
+	payload, ok := s.pendingPayloads[req.PayloadId]
```
need to protect `s.pendingPayload` with read-write lock or something

yperbasis commented on cmd/rpcdaemon/commands/engine_api.go:
```diff
@@ -2,41 +2,35 @@ package commands
+	BaseFeePerGas *hexutil.Big    `json:"baseFeePerGas"`
+	BlockHash     common.Hash     `json:"blockHash"`
+	Timestamp     hexutil.Uint64  `json:"timestamp"`
+	Coinbase      common.Address  `json:"coinbase"`
```
Should be called "feeRecipient" instead per the [Engine API](https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#executionpayloadv1).

yperbasis replied:
Ah, actually no, sorry, it's "coinbase" in [v1.0.0-alpha.4](https://github.com/ethereum/execution-apis/blob/v1.0.0-alpha.4/src/engine/specification.md#executionpayloadv1).

yperbasis commented on cmd/rpcdaemon/commands/engine_api.go:
```diff
@@ -2,41 +2,35 @@ package commands
-	Timestamp     *hexutil.Uint64        `json:"timestamp"`
-	Coinbase      *common.Address        `json:"coinbase"`
-	Transactions  *[]hexutil.Bytes       `json:"transactions"`
+	ParentHash    common.Hash     `json:"parentHash"`
```
We should probably slap `gencodec:"required"` on all fields.

yperbasis commented on cmd/rpcdaemon/commands/engine_api.go:
```diff
@@ -46,59 +40,121 @@ type PreparePayloadArgs struct {
 	FeeRecipient *common.Address `json:"feeRecipients"`
```
Should be `json:"feeRecipient"`, not `json:"feeRecipients"`

yperbasis commented on cmd/rpcdaemon/commands/engine_api.go:
```diff
@@ -2,103 +2,159 @@ package commands
 		}, nil
 	}
+	// Request for assembling payload
+	return nil, nil
```
We should return an error indicating that we can't build payloads yet. Probably -32000 "Server error".

yperbasis commented on cmd/rpcdaemon/commands/engine_api.go:
```diff
@@ -2,103 +2,159 @@ package commands
+func (e *EngineImpl) ExecutePayloadV1(ctx context.Context, payload ExecutionPayload) (map[string]interface{}, error) {
+	var baseFee *uint256.Int
+	if payload.BaseFeePerGas != nil {
+		baseFee, _ = uint256.FromBig((*big.Int)(payload.BaseFeePerGas))
```
We should return an error (perhaps -32600 "Invalid Request") in case payload.BaseFeePerGas overflows 256 bits.

yperbasis commented on ethdb/privateapi/ethbackend.go:
```diff
@@ -38,8 +60,20 @@ type EthBackend interface {
+// Status: block's status
+type ExecutionStatus struct {
+	HeadHash common.Hash
+	Status   string
```
Better use an enum

yperbasis commented on eth/backend.go:
```diff
@@ -358,13 +365,15 @@ func New(stack *node.Node, config *ethconfig.Config, logger log.Logger) (*Ethere
 		ethashApi = casted.APIs(nil)[1].Service.(*ethash.API)
 	}
-	ethBackendRPC := privateapi.NewEthBackendServer(ctx, backend, backend.chainDB, backend.notifications.Events, blockReader)
+	backend.waitingForPOSHeaders = false
```
Is there any piece of code that sets `waitingForPOSHeaders` to true?

Giulio2002 replied:
It will be used in downward sync, in future PRs

yperbasis commented on eth/stagedsync/stage_headers.go:
```diff
@@ -57,6 +57,50 @@ func StageHeadersCfg(
 	}
 }
 
+func SpawnStageHeaders(
```
`SpawnStageHeaders` isn't used anywhere, is it?

Giulio2002 replied:
I now call it, however it does nothing, it is just a starting point for downward download

yperbasis replied:
Hmm. It looks like using `SpawnStageHeaders` in https://github.com/ledgerwatch/erigon/pull/2997/commits/19785b42c46c19318fe696ecbedbf523a596af6a is causing the tests to fail.

I'd suggest to remove `SpawnStageHeaders` for now to reduce the scope of this PR.

Giulio2002 replied:
okay removed it for now

yperbasis commented on ethdb/privateapi/engine_test.go:
```diff
@@ -0,0 +1,305 @@
+	done := make(chan bool)
+
+	go func() {
+		_, err = backend.EngineExecutePayloadV1(ctx, &types2.ExecutionPayload{
```
Why is this payload invalid? It's not immediately clear. Please add a clarifying comment.

Giulio2002 replied:
Ok, so we are giving status invalid, this is meant to be done in stage execution, if that block gets executed with no error we give off valid if not invalid.

yperbasis replied:
Sorry, I'm not following. Is there some special logic somewhere to return invalid status for this test, or the payload specified in this test is somehow invalid and will be rejected due to a standard validity check?

Giulio2002 replied:
comment added

yperbasis replied:
In other words, why are we expecting "invalid" status?

Giulio2002 replied:
there is no special logic, Invalid is returned in case we have a problem executing the block for some reason. so invalid is returned if the block is a bad block

Giulio2002 replied:
in the test we mock the behaviour that the execute stage will have towards an invalid payload

Giulio2002 replied:
we are expecting invalid status because that's what the statusCh sent us, that will be supposed to happen in stage execution

yperbasis replied:
Sorry, I'm still not following. Where is the piece of code that sends invalid status to `statusCh` for `TestInvalidRequest`? Does it do because it was mocked to always return invalid status or because the payload in `TestInvalidRequest` is somehow invalid?

Giulio2002 replied:
it was mocked, `statusCh <- ExecutionStatus{Status: Invalid}`is the mock

Giulio2002 replied:
statusCh is supposed to be updated in stage execution where we can determine wheter the block is good or bad

Giulio2002 replied:
the payload itself is not invalid in itself

yperbasis replied:
I see `statusCh <- ExecutionStatus{Status: Invalid}` in `TestMockInvalidExecution`, but not in `TestInvalidRequest`. So my question stands.

Giulio2002 replied:
TestInvalidRequest is if we give a request that is not valid, like a malformed block. for example if it misses some fields, since they are all defined as pointers in gointerface

yperbasis replied:
All right.  So please put a comment into `TestInvalidRequest` clarifying that the request is invalid because field XXX is missing.

Giulio2002 replied:
i made it more clear now, sorry

yperbasis commented:
Here's a new test failure to fix:
```
--- FAIL: TestMockDownloadRequest (0.00s)
    engine_test.go:116: 
        	Error Trace:	engine_test.go:116
        	Error:      	Not equal: 
        	            	expected: string("SYNCING")
        	            	actual  : privateapi.PayloadStatus("SYNCING")
        	Test:       	TestMockDownloadRequest
```

Giulio2002 commented:
fixed

> Here's a new test failure to fix:
> 
> ```
> --- FAIL: TestMockDownloadRequest (0.00s)
>     engine_test.go:116: 
>         	Error Trace:	engine_test.go:116
>         	Error:      	Not equal: 
>         	            	expected: string("SYNCING")
>         	            	actual  : privateapi.PayloadStatus("SYNCING")
>         	Test:       	TestMockDownloadRequest
> ```

fixed it into 8ee29ee

yperbasis reviewed (approved):
We'll probably have to revisit various bits while we keep working on the Merge, but is probably fine as a stepping stone.
